python/graphstorm/model/utils.py

```diff
@@ -703,8 +703,109 @@ def shuffle_predict(predictions, id_mapping_file, pred_type,
                 len(predictions)).cpu() # predictions are stored in CPU
     return predictions[local_id_mapping]
 
+class NodeIDShuffler():
+    """ Shuffle node ids into the original node ids according to node_id_mappings
+
+        Parameters
+        ----------
+        g: Dist DGLGraph
+            Graph.
+        node_id_mapping_file:
+            Path to the file storing node id mapping generated by the
+            graph partition algorithm.
+    """
+    def __init__(self, g, node_id_mapping_file):
+        self._g = g
+        assert node_id_mapping_file is not None \
+            and os.path.exists(node_id_mapping_file), \
+            f"{node_id_mapping_file} must exist."
+
+        # _id_mapping_info stores node ids in distributed tensors.
+        # ID mappings are not load in __init__ function
+        # The mappings are loaded on-demand.
+        self._id_mapping_info = {}
+        # _id_mappings is used by rank0 to cache the node id mapping file.
+        if get_rank() == 0:
+            self._id_mappings = th.load(node_id_mapping_file)
```

Comment:
this is saved in rank 0 process and the data are stored here whenever the object exists. should we save the data in `self._id_mapping_info` directly and avoid storing `self._id_mappings`?

Reply:
Storing data in self._id_mapping_info requires storing data as Dist Tensors, i.e., on shared memory, while holding self._id_mappings only consumes non shared memory.
This implementation try to reduce the overall cost of shared memory, as in some cases, not all the node id mappings are required.

Reply:
> this is saved in rank 0 process and the data are stored here whenever the object exists. should we save the data in `self._id_mapping_info` directly and avoid storing `self._id_mappings`?

Storing data in self._id_mapping_info requires storing data as Dist Tensors, i.e., on shared memory, while holding self._id_mappings only consumes non shared memory.
This implementation try to reduce the overall cost of shared memory, as in some cases, not all the node id mappings are required.

Reply:
at the cost of having one process to store the data of all nodes? i'm not sure if it will make it more scalable.

Reply:
i see some benefits if there is only one machine. otherwise, storing data in shared memory should be better.

Reply:
Updated